lib/dbus/bus.rb

```diff
@@ -538,7 +538,8 @@ def process(m)
         end
       when DBus::Message::SIGNAL
         # the signal can match multiple different rules
-        @signal_matchrules.each do |mrs, slot|
+        # clone to allow new signale handlers to be registered
+        @signal_matchrules.dup.each do |mrs, slot|
```

Comment:
OK! I suppose that the condition that this avoids is the "RuntimeError: can't add a new key into hash during iteration" exception, right?
Can you please quote some code that you use to exercise this, so that I can add a test case? Or even better, add a test case yourself :)

Reply:
> OK! I suppose that the condition that this avoids is the "RuntimeError: can't add a new key into hash during iteration" exception, right?

Exactly. :) We're waiting for a service to appear at the bus and then instantly register a signal handler. The condensed code looks like this:

```ruby
@dbus = DBus::ASessionBus.new
@service_names = [ 'bla', 'blub' ]

@dbus.proxy.on_signal('NameOwnerChanged') do |sname, old_owner, new_owner|
  next unless @service_names.include?(sname)

  if new_owner.present? && !old_owner.present?
    Rails.logger.info { "DBusChange: detected service #{sname} start #{[ sname, old_owner, new_owner ].inspect}" }

    if !@service_proxy && (service = @dbus.service(sname)) && service.exists?
      begin
        client = service.object(Service::INSTANCE)
        client.introspect
        client.default_iface = Service::INTERFACE
        @service_proxy = client

        @service_proxy.on_signal('my_signal') do |param|
          Rails.logger.info { "service #{sname} sent signal #{'my_signal'}" }
        end
      rescue DBus::Error, RuntimeError => e
        Rails.logger.error { "ERROR: #{e.inspect}" }
      end
    end
  end
end
```